src/AspNetCoreModuleV2/CommonLib/hostfxr_utility.cpp

```diff
@@ -394,163 +310,85 @@ HOSTFXR_UTILITY::ParseHostfxrArguments(
+    auto hostFxrBase = dotnetPath.parent_path() / L"host\\fxr";
 
-    if (FAILED(hr = struHostFxrPath.Copy(pStruAbsolutePathToDotnet)))
+    if (!is_directory(hostFxrBase))
     {
-        goto Finished;
-    }
-
-    dwPosition = struHostFxrPath.LastIndexOf(L'\\', 0);
-    if (dwPosition == -1)
-    {
-        hr = E_FAIL;
-        goto Finished;
+        EVENTLOG(hEventLog, HOSTFXR_DIRECTORY_NOT_FOUND, hostFxrBase.c_str(), HRESULT_FROM_WIN32(ERROR_BAD_ENVIRONMENT));
+        return std::nullopt;
     }
 
-    struHostFxrPath.QueryStr()[dwPosition] = L'\0';
-
-    if (FAILED(hr = struHostFxrPath.SyncWithBuffer()) ||
-        FAILED(hr = struHostFxrPath.Append(L"\\")))
-    {
-        goto Finished;
-    }
-
-    hr = struHostFxrPath.Append(L"host\\fxr");
-    if (FAILED(hr))
-    {
-        goto Finished;
-    }
-
-    if (!UTILITY::DirectoryExists(&struHostFxrPath))
-    {
-        hr = ERROR_BAD_ENVIRONMENT;
-        UTILITY::LogEventF(hEventLog,
-            EVENTLOG_ERROR_TYPE,
-            ASPNETCORE_EVENT_HOSTFXR_DIRECTORY_NOT_FOUND,
-            struEventMsg.QueryStr(),
-            ASPNETCORE_EVENT_HOSTFXR_DIRECTORY_NOT_FOUND_MSG,
-            struHostFxrPath.QueryStr(),
-            hr);
-        goto Finished;
-    }
-
-    // Find all folders under host\\fxr\\ for version numbers.
-    hr = struHostFxrSearchExpression.Copy(struHostFxrPath);
-    if (FAILED(hr))
-    {
-        goto Finished;
-    }
-
-    hr = struHostFxrSearchExpression.Append(L"\\*");
-    if (FAILED(hr))
-    {
-        goto Finished;
-    }
+    auto searchPattern = std::wstring(hostFxrBase) + L"\\*";
 
     // As we use the logic from core-setup, we are opting to use std here.
-    HOSTFXR_UTILITY::FindDotNetFolders(struHostFxrSearchExpression.QueryStr(), &vVersionFolders);
+    FindDotNetFolders(searchPattern.c_str(), vVersionFolders);
 
-    if (vVersionFolders.size() == 0)
+    if (vVersionFolders.empty())
     {
-        hr = HRESULT_FROM_WIN32(ERROR_BAD_ENVIRONMENT);
-        UTILITY::LogEventF(hEventLog,
-            EVENTLOG_ERROR_TYPE,
-            ASPNETCORE_EVENT_HOSTFXR_DIRECTORY_NOT_FOUND,
-            ASPNETCORE_EVENT_HOSTFXR_DIRECTORY_NOT_FOUND_MSG,
-            struHostFxrPath.QueryStr(),
-            hr);
-        goto Finished;
+        EVENTLOG(hEventLog, HOSTFXR_DIRECTORY_NOT_FOUND, hostFxrBase.c_str(), HRESULT_FROM_WIN32(ERROR_BAD_ENVIRONMENT));
+        return std::nullopt;
     }
 
-    hr = FindHighestDotNetVersion(vVersionFolders, &struHighestDotnetVersion);
-    if (FAILED(hr))
-    {
-        goto Finished;
-    }
+    const auto highestVersion = FindHighestDotNetVersion(vVersionFolders);
+    const auto hostFxrPath = hostFxrBase / highestVersion / "hostfxr.dll";
 
-    if (FAILED(hr = struHostFxrPath.Append(L"\\"))
-        || FAILED(hr = struHostFxrPath.Append(struHighestDotnetVersion.QueryStr()))
-        || FAILED(hr = struHostFxrPath.Append(L"\\hostfxr.dll")))
+    if (!is_regular_file(hostFxrPath))
```

Comment:
fs::is_regular_file ?

Reply:
why?

Reply:
is_regular_file is fine, but exists(path) was a bit harder for me to understand. Didn't know where the exists method was coming from.